File: libs/core/src/symbols/merged_namespace_symbol.cpp
/**
 * @file merged_namespace_symbol.cpp
 * @author Francesco Corso
 * @date 8/5/2026
 * @brief
 */
module;

#include <libassert/assert-macros.hpp>

module prism.core:symbols.merged_namespace_symbol.impl;

import :symbols.merged_namespace_symbol;
import :compilation;
import :symbols.visit;

namespace prism
{

    const NamespaceSymbol &MergedNamespaceSymbol::create(const Compilation &compilation,
                                                         const NamespaceSymbol *containing_namespace,
                                                         std::vector<Ref<const NamespaceSymbol>> namespaces)
    {
        DEBUG_ASSERT(!namespaces.empty(), "Cannot create merged namespace symbol with no namespaces");

        if (namespaces.size() == 1)
            return namespaces.front();

        auto &lifetime = CompilationInternal::get_lifetime(compilation);
        auto name = namespaces.front()->name();
        return lifetime.create<MergedNamespaceSymbol>(name, containing_namespace, compilation, std::move(namespaces));
    }

    const ImmutableArray<Location> &MergedNamespaceSymbol::locations() const
    {
        return locations_.get_or_compute(
            [this]
            {
                return namespaces_ |
                       std::views::transform([](const NamespaceSymbol &n) -> auto & { return n.locations(); }) |
                       std::views::join | std::ranges::to<ImmutableArray>();
            });
    }

    std::span<const SyntaxReference> MergedNamespaceSymbol::declaring_syntax_references() const
    {
        return syntax_references_.get_or_compute(
            [this]
            {
                return namespaces_ |
                       std::views::transform([](const NamespaceSymbol &n) { return n.declaring_syntax_references(); }) |
                       std::views::join | std::ranges::to<std::vector>();
            });
    }

    Optional<const Compilation &> MergedNamespaceSymbol::containing_compilation() const noexcept
    {
        return compilation_;
    }

    SymbolSpan<Symbol> MergedNamespaceSymbol::members() const
    {
        return members_.get_or_compute(
            [this]
            {
                std::unordered_map<Name, std::vector<Ref<const NamespaceSymbol>>> namespace_groups;

                for (const auto sym : namespaces_ |
                                          std::views::transform([](const NamespaceSymbol &n) { return n.members(); }) |
                                          std::views::join)
                {
                    if (auto ns = sym->as<NamespaceSymbol>(); ns.has_value())
                    {
                        namespace_groups[sym->name()].emplace_back(*ns);
                    }
                }

                std::unordered_set<Name> emitted_namespaces;
                std::vector<Ref<const Symbol>> merged_members;
                for (const auto member :
                     namespaces_ | std::views::transform([](const NamespaceSymbol &n) { return n.members(); }) |
                         std::views::join)
                {
                    if (auto ns = member->as<NamespaceSymbol>(); ns.has_value())
                    {
                        if (emitted_namespaces.insert(member->name()).second)
                        {
                            merged_members.emplace_back(
                                create(compilation_, this, std::move(namespace_groups[member->name()])));
                        }

                        continue;
                    }

                    merged_members.emplace_back(member);
                }

                return merged_members;
            });
    }

    SymbolSpan<Symbol> MergedNamespaceSymbol::members(Name name) const
    {
        Lazy<ImmutableArray<Ref<const Symbol>>> *lazy_array;
        {
            std::scoped_lock lock{name_to_members_mutex_};
            lazy_array = &name_to_members_[name];
        }

        return lazy_array->get_or_compute([this, name] { return compute_members(name); });
    }

    void MergedNamespaceSymbol::force_complete(const Optional<SourceLocation> &location,
                                               const Optional<SymbolPredicate> &filter) const
    {
        for (auto part : namespaces_)
        {
            SymbolInternal::force_complete(part, location, filter);
        }
    }

    ImmutableArray<Ref<const Symbol>> MergedNamespaceSymbol::compute_members(const Name name) const
    {
        return namespaces_ | std::views::transform([&](const NamespaceSymbol &n) { return n.members(name); }) |
               std::views::join | std::ranges::to<ImmutableArray>();
    }
} // namespace prism

File: src/Prism.Core/Symbols/MergedNamespaceSymbol.cs
using System.Collections.Concurrent;
using System.Collections.Immutable;
using System.Diagnostics;
using Prism.Core.Compiling;
using Prism.Core.Diagnostics;
using Prism.Core.Syntax;
using ZLinq;

namespace Prism.Core.Symbols;

internal sealed class MergedNamespaceSymbol : NamespaceSymbol
{
    private readonly ImmutableArray<NamespaceSymbol> _namespaces;
    private ImmutableArray<Symbol> _members;
    private readonly ConcurrentDictionary<string, ImmutableArray<Symbol>> _nameToMembers = new();

    public MergedNamespaceSymbol(
        string name,
        Symbol? containingSymbol,
        Compilation compilation,
        ImmutableArray<NamespaceSymbol> namespaces
    )
        : base(name, containingSymbol)
    {
        ContainingCompilation = compilation;
        _namespaces = namespaces;
    }

    public static NamespaceSymbol Create(
        Compilation compilation,
        NamespaceSymbol? containingSymbol,
        ImmutableArray<NamespaceSymbol> namespaces
    )
    {
        Debug.Assert(!namespaces.IsDefaultOrEmpty, "Cannot merge an empty set of namespaces");

        if (namespaces.Length == 1)
            return namespaces[0];

        return new MergedNamespaceSymbol(
            namespaces[0].Name,
            containingSymbol,
            compilation,
            namespaces
        );
    }

    public override ImmutableArray<Location> Locations
    {
        get
        {
            if (!field.IsDefault)
                return field;

            Interlocked.CompareExchange(
                ref field,
                [
                    .. _namespaces
                        .AsValueEnumerable()
                        .SelectMany(x => x.Locations.AsValueEnumerable()),
                ],
                default
            );
            return field;
        }
    }

    public override ImmutableArray<SyntaxReference> DeclaringSyntaxReferences
    {
        get
        {
            if (!field.IsDefault)
                return field;

            Interlocked.CompareExchange(
                ref field,
                [
                    .. _namespaces
                        .AsValueEnumerable()
                        .SelectMany(x => x.DeclaringSyntaxReferences.AsValueEnumerable()),
                ],
                default
            );
            return field;
        }
    }

    public override ImmutableArray<Symbol> GetMembers()
    {
        if (!_members.IsDefault)
            return _members;

        var namespaceGroups = _namespaces
            .AsValueEnumerable()
            .SelectMany(x => x.GetMembers().AsValueEnumerable())
            .OfType<NamespaceSymbol>()
            .GroupBy(x => x.Name)
            .ToDictionary(x => x.Key, x => x.ToImmutableArray());

        var emittedNamespaces = new HashSet<string>();
        var mergedMembers = ImmutableArray.CreateBuilder<Symbol>();

        foreach (
            var member in _namespaces
                .AsValueEnumerable()
                .SelectMany(x => x.GetMembers().AsValueEnumerable())
        )
        {
            if (member is NamespaceSymbol ns)
            {
                if (emittedNamespaces.Add(ns.Name))
                {
                    mergedMembers.Add(
                        Create(ContainingCompilation, this, namespaceGroups[ns.Name])
                    );
                }

                continue;
            }

            mergedMembers.Add(member);
        }

        Interlocked.CompareExchange(ref _members, mergedMembers.DrainToImmutable(), default);
        return _members;
    }

    public override ImmutableArray<Symbol> GetMembers(string name)
    {
        return _nameToMembers.GetOrAdd(name, static (n, self) => self.ComputeMembers(n), this);
    }

    private ImmutableArray<Symbol> ComputeMembers(string name)
    {
        return
        [
            .. _namespaces
                .AsValueEnumerable()
                .SelectMany(x => x.GetMembers(name).AsValueEnumerable()),
        ];
    }

    public override NamespaceKind NamespaceKind => NamespaceKind.Compilation;
    public override Compilation ContainingCompilation { get; }

    internal override void ForceComplete(
        SourceLocation? location,
        Predicate<Symbol>? filter,
        CancellationToken cancellationToken
    )
    {
        foreach (var ns in _namespaces)
        {
            ns.ForceComplete(location, filter, cancellationToken);
        }
    }
}
